quinn-proto/src/crypto.rs

```diff
@@ -124,3 +126,16 @@ pub trait HeaderKeys {
     /// The sample size used for this key's algorithm
     fn sample_size(&self) -> usize;
 }
+
+/// A key for signing with HMAC-based algorithms
+pub trait HmacKey {
+    /// Type of the signatures created by `sign()`
+    type Signature: AsRef<[u8]>;
+
+    /// Method for creating a key
+    fn new(key: &[u8]) -> Self;
+    /// Method for signing a message
+    fn sign(&self, data: &[u8]) -> Self::Signature;
+    /// Method for verifying a message
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<(), ()>;
+}
```

Comment:
Could this have a `const KEY_SIZE: usize` (bearing in mind that associated constants aren't usable in array sizes yet, so allocations would still have to be dynamic) or similar so we don't have to inflexibly hardcode 64 bytes everywhere?

Reply:
I have that mostly done, but then tried to figure out the thing about associated constants as array sizes. Will push it later.

Reply:
Yeah, you'll probably have to switch over to `Vec` everywhere, which I think is fine.

Reply:
So what do you propose slices of an incorrect length do? Panic, return an `Err`? I kind of feel like pending array sizes that actually allow using associated consts and pending an actual other implementation of the trait, keeping the fixed size is actually more ergonomic. (Also arguably I feel like this shouldn't prevent us from landing the rest of these changes which clearly make the API more flexible either way.)

Reply:
> So what do you propose slices of an incorrect length do?

I think we can already return `ConfigError` from every entry point that consumes a config struct; this is not the first case of a config option that can have incoherent values, and this case does not strike me as significantly different.

> I feel like this shouldn't prevent us from landing the rest of these changes

A fixed key size excludes the majority of reasonable implementations. I agree that it's not a hard blocker, but it is a significant wart.

Reply:
Fixed.